### PR TITLE
Implement iter::size_hint and ExactSizeIterator for Witness Iter

### PR DESCRIPTION
**Pull request by RCasatta**
close https://github.com/rust-bitcoin/rust-bitcoin/issues/1050

I don't think we need to change the `collect` since it use the `size_hint()` lower bound to initially allocate

// with size_hint
// test blockdata::witness::benches::bench_big_witness_to_vec              ... bench:         313 ns/iter (+/- 13)
// test blockdata::witness::benches::bench_witness_to_vec                  ... bench:         204 ns/iter (+/- 11)

// without
// test blockdata::witness::benches::bench_big_witness_to_vec              ... bench:         489 ns/iter (+/- 28)
// test blockdata::witness::benches::bench_witness_to_vec                  ... bench:         221 ns/iter (+/- 102)

The reason why the small witness doesn't get big perf boost is because by default vec allocates 4 slots

### CODE REVIEW

File: src/blockdata/witness.rs [hunk 1]
```diff
@@ -43,7 +43,7 @@ pub struct Witness {
 }
 
 /// Support structure to allow efficient and convenient iteration over the Witness elements
-pub struct Iter<'a>(::core::slice::Iter<'a, u8>);
+pub struct Iter<'a>(::core::slice::Iter<'a, u8>, usize);
```
Comment (dpc) [inline src/blockdata/witness.rs:46]: If it is already two fields, I would name them.

**Review by dpc — APPROVED** [review]

**Review by tcharding — APPROVED** [review]
ACK 52014d740f19f2b8d16d71bbc4363598b3c1a2e0

File: src/blockdata/witness.rs [hunk 2]
```diff
@@ -43,7 +43,10 @@ pub struct Witness {
 }
 
 /// Support structure to allow efficient and convenient iteration over the Witness elements
-pub struct Iter<'a>(::core::slice::Iter<'a, u8>);
+pub struct Iter<'a> {
+    inner: ::core::slice::Iter<'a, u8>,
```
Comment (tcharding) [inline src/blockdata/witness.rs:47]: nit: We don't need these leading colons anymore, since the MSRV bump.

**Review by apoelstra — APPROVED** [review]
ACK 52014d740f19f2b8d16d71bbc4363598b3c1a2e0 nice!

**Review by Kixunil — DISMISSED** [review]
LOL at first I thought small ones got worse, I'm actually surprised they didn't. (Would be happy to see an explanation if anyone knows.)

ACK 52014d740f19f2b8d16d71bbc4363598b3c1a2e0

File: src/blockdata/witness.rs (hunk 2 above)
Comment (RCasatta) [reply src/blockdata/witness.rs]: fixed

**Review by Kixunil — APPROVED** [review]
ACK ec8dadaf86b23c4bf40376abc3b4b8a086e9fdf4

**Review by apoelstra — APPROVED** [review]
ACK ec8dadaf86b23c4bf40376abc3b4b8a086e9fdf4

**Pull request merged**